src/Microsoft.OData.Client/BulkUpdateSaveResult.cs

```diff
@@ -202,19 +275,350 @@ protected override ODataRequestMessageWrapper CreateRequestMessage(string method
+                        {
+                            if (this.adjacentDescriptors.Peek().Item1 == this.adjacentDescriptors.Peek().Item2.Count - 1)
+                            {
+                                this.adjacentDescriptors.Pop();
+                            }
+
+                            int index = this.adjacentDescriptors.Peek().Item1;
+                            IReadOnlyList<Descriptor> descriptors = this.adjacentDescriptors.Peek().Item2;
+
+                            if (index == descriptors.Count - 1)
+                            {
+                                this.adjacentDescriptors.Pop();
+                            }
+                            else
+                            {
+                                descriptor = descriptors[index + 1];
+                                this.adjacentDescriptors.Pop();
+                                this.adjacentDescriptors.Push((index + 1, descriptors));
+                            }
+                        }
+                        else 
+                        {
+                            descriptor = relatedDescriptors[0];
+                            this.adjacentDescriptors.Push((0, relatedDescriptors));
+                        }
+                    }          
+                }
+                 
+                if (descriptor != null) 
+                {
+                    this.SaveResultProcessed(descriptor, HasFailedOperation(materializerState));
+
+                    operationResponse = new ChangeOperationResponse(this.cachedResponse.Headers, descriptor)
+                    {
+                        StatusCode = (int)this.cachedResponse.StatusCode
+                    };
+
+                    this.materializerStateForDescriptor.Add(descriptor, materializerState);
+
+                    if (this.cachedResponse.Exception != null)
+                    {
+                        operationResponse.Error = this.cachedResponse.Exception;
+
+                        if (ex == null)
+                        {
+                            ex = this.cachedResponse.Exception;
+                        }
+                    }
+                    else
+                    {
+#if DEBUG
+                        this.HandleOperationResponse(descriptor, this.cachedResponse.Headers, this.cachedResponse.StatusCode);
+#else
+                        this.HandleOperationResponse(descriptor, this.cachedResponse.Headers);
+#endif
+                    }
+
+                    if (isTopLevelResponse)
+                    {
+                        this.operationResponses.Add(operationResponse);
+                        this.operationResponseTracker.Push(this.operationResponses);
+                    }
+                    else
+                    {
+                        List<OperationResponse> lastOperationResponses = this.operationResponseTracker.Peek();
+                        lastOperationResponses.Last().NestedResponses.Add(operationResponse);

```

Comment:
You can avoid using the expensive LINQ `Last()` method since you're working with a `List<T>`

Reply:
How do I get the last item ?

Reply:
@ElizabethOkerio Since `lastOperationResponses` is a list, `lastOperationResponses[lastOperationResponses.Count - 1]` is less expensive than `lastOperationResponses.Last()`.